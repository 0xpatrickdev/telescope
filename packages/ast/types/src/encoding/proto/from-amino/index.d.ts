import * as t from '@babel/types';
import { ProtoField } from '@cosmology/types';
import { ProtoParseContext } from '../../context';
export interface FromAminoJSONMethod {
    context: ProtoParseContext;
    field: ProtoField;
    isOptional: boolean;
}
export declare const fromAminoJSONMethodFields: (context: ProtoParseContext, name: string, proto: ProtoType) => t.ObjectProperty[];
export declare const fromAminoJSONMethod: (context: ProtoParseContext, name: string, proto: ProtoType) => t.ObjectMethod;
export declare const fromAminoMsgMethod: (context: ProtoParseContext, name: string, proto: ProtoType) => t.ObjectMethod;
