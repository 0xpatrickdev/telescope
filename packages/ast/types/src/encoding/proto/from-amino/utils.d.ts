import * as t from '@babel/types';
import { FromAminoJSONMethod } from './index';
export declare const fromAminoJSON: {
    scalar(args: FromAminoJSONMethod): t.ObjectProperty;
    string(args: FromAminoJSONMethod): t.ObjectProperty;
    number(args: FromAminoJSONMethod): t.ObjectProperty;
    double(args: FromAminoJSONMethod): t.ObjectProperty;
    float(args: FromAminoJSONMethod): t.ObjectProperty;
    int32(args: FromAminoJSONMethod): t.ObjectProperty;
    sint32(args: FromAminoJSONMethod): t.ObjectProperty;
    uint32(args: FromAminoJSONMethod): t.ObjectProperty;
    fixed32(args: FromAminoJSONMethod): t.ObjectProperty;
    sfixed32(args: FromAminoJSONMethod): t.ObjectProperty;
    bool(args: FromAminoJSONMethod): t.ObjectProperty;
    long(args: FromAminoJSONMethod): t.ObjectProperty;
    int64(args: FromAminoJSONMethod): t.ObjectProperty;
    uint64(args: FromAminoJSONMethod): t.ObjectProperty;
    sint64(args: FromAminoJSONMethod): t.ObjectProperty;
    fixed64(args: FromAminoJSONMethod): t.ObjectProperty;
    sfixed64(args: FromAminoJSONMethod): t.ObjectProperty;
    type(args: FromAminoJSONMethod): t.ObjectProperty;
    enum(args: FromAminoJSONMethod): t.ObjectProperty;
    bytes(args: FromAminoJSONMethod): t.ObjectProperty;
    duration(args: FromAminoJSONMethod): t.ObjectProperty;
    durationString(args: FromAminoJSONMethod): t.ObjectProperty;
    timestamp(args: FromAminoJSONMethod): t.ObjectProperty;
    keyHash(args: FromAminoJSONMethod): t.ObjectProperty;
    array(args: FromAminoJSONMethod, expr: t.Expression): t.ObjectProperty;
};
export declare const arrayTypes: {
    scalar(): t.Identifier;
    string(): t.Identifier;
    bool(): t.Identifier;
    bytes(args: FromAminoJSONMethod): t.Identifier;
    long(): t.Identifier;
    uint64(): t.Identifier;
    int64(): t.Identifier;
    sint64(): t.Identifier;
    fixed64(): t.Identifier;
    sfixed64(): t.Identifier;
    number(): t.Identifier;
    uint32(): t.Identifier;
    int32(): t.Identifier;
    sint32(): t.Identifier;
    fixed32(): t.Identifier;
    sfixed32(): t.Identifier;
    double(): t.Identifier;
    float(): t.Identifier;
    enum(args: FromAminoJSONMethod): t.CallExpression;
    type(args: FromAminoJSONMethod): t.CallExpression;
};
