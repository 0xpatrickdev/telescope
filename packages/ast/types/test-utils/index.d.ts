import { ProtoType } from '@osmonauts/types';
import { ProtoStore } from '@osmonauts/proto-parser';
import { AminoParseContext } from '../src/encoding/context';
export declare const expectCode: (ast: any) => void;
export declare const printCode: (ast: any) => void;
export declare const prepareContext: (store: ProtoStore, protoFile: string) => {
    context: AminoParseContext;
    root: import("@osmonauts/types").ProtoRoot;
    protos: ProtoType[];
};
