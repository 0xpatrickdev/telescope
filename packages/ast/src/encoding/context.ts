import { ProtoStore, ProtoRef } from '@osmonauts/proto-parser';

export interface ParseContext {
    utils: Record<string, boolean>;
    addUtil: Function;
}

export class GenericParseContext implements ParseContext {
    imports: Record<string, object> = {};
    utils: Record<string, boolean> = {};

    addUtil(util) {
        this.utils[util] = true;
    }
}
export class AminoParseContext implements ParseContext {
    store: ProtoStore;
    ref: ProtoRef;

    utils: Record<string, boolean> = {};

    constructor(
        ref: ProtoRef,
        store: ProtoStore
    ) {
        this.ref = ref;
        this.store = store;
    }

    addUtil(util) {
        this.utils[util] = true;
    }
}

export class ProtoParseContext implements ParseContext {
    imports: Record<string, object> = {};
    utils: Record<string, boolean> = {};

    addUtil(util) {
        this.utils[util] = true;
    }
}

