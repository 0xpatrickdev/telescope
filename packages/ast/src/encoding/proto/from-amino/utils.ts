import * as t from "@babel/types";
import { fromAminoMethod } from "./index";
import { identifier, callExpression, TypeLong, BILLION } from "../../../utils";
import { getFieldsTypeName } from "..";
import { getDefaultTSTypeFromProtoType, getFieldNames } from "../../types";
import { ProtoType } from "@cosmology/types";
import { ProtoParseContext } from "../../context";
import { camel } from "@cosmology/utils";

// message.sender = expr
const setField = (args: fromAminoMethod, expr: t.Expression): t.Statement => {
  const { propName } = getFieldNames(args.field);

  return t.expressionStatement(
    t.assignmentExpression(
      "=",
      t.memberExpression(t.identifier("message"), t.identifier(propName)),
      expr
    )
  );
};

const setNullishCoalescing = (
  args: fromAminoMethod,
  value: t.Expression
): t.Statement => {
  const { origName } = getFieldNames(args.field);
  return setField(
    args,
    t.logicalExpression(
      "??",
      t.memberExpression(t.identifier("object"), t.identifier(origName)),
      value
    )
  );
};

const setNotUndefinedAndNotNull = (
  value: t.Expression,
  defaultValue: t.Expression,
  args: fromAminoMethod
): t.Statement => {
  const { propName, origName } = getFieldNames(args.field);

  return t.ifStatement(
    t.logicalExpression(
      "&&",
      t.binaryExpression(
        "!==",
        t.memberExpression(t.identifier("object"), t.identifier(origName)),
        t.identifier("undefined")
      ),
      t.binaryExpression(
        "!==",
        t.memberExpression(t.identifier("object"), t.identifier(origName)),
        t.nullLiteral()
      )
    ),
    t.blockStatement([
      t.expressionStatement(
        t.assignmentExpression(
          "=",
          t.memberExpression(t.identifier("message"), t.identifier(propName)),
          value
        )
      ),
    ])
  );
};

export const fromAmino = {
  // message.sender = object.sender ?? "";
  string(args: fromAminoMethod) {
    return setNullishCoalescing(
      args,
      getDefaultTSTypeFromProtoType(args.context, args.field, args.isOneOf)
    );
  },

  // message.disableMacros = object.disableMacros ?? false;
  bool(args: fromAminoMethod) {
    return setNullishCoalescing(
      args,
      getDefaultTSTypeFromProtoType(args.context, args.field, args.isOneOf)
    );
  },

  // message.doubleValue = object.doubleValue ?? 0;

  number(args: fromAminoMethod) {
    return setNullishCoalescing(
      args,
      getDefaultTSTypeFromProtoType(args.context, args.field, args.isOneOf)
    );
  },

  int32(args: fromAminoMethod) {
    return fromAmino.number(args);
  },
  uint32(args: fromAminoMethod) {
    return fromAmino.number(args);
  },
  sint32(args: fromAminoMethod) {
    return fromAmino.number(args);
  },
  fixed32(args: fromAminoMethod) {
    return fromAmino.number(args);
  },
  sfixed32(args: fromAminoMethod) {
    return fromAmino.number(args);
  },
  double(args: fromAminoMethod) {
    return fromAmino.number(args);
  },
  float(args: fromAminoMethod) {
    return fromAmino.number(args);
  },

  // OLD: message.myInt64Value = object.myInt64Value !== undefined && object.myInt64Value !== null ? Long.fromValue(object.myInt64Value) : Long.ZERO;
  // NEW: if( object.myInt64Value !== undefined && object.myInt64Value !== null ) { message.myInt64Value = Long.fromValue(object.myInt64Value) }
  long(args: fromAminoMethod) {
    const { origName } = getFieldNames(args.field);

    TypeLong.addUtil(args.context);

    return setNotUndefinedAndNotNull(
      TypeLong.getFromValueWithArgs(
        args.context,
        t.memberExpression(t.identifier("object"), t.identifier(origName))
      ),
      getDefaultTSTypeFromProtoType(args.context, args.field, args.isOneOf),
      args
    );
  },

  int64(args: fromAminoMethod) {
    return fromAmino.long(args);
  },
  uint64(args: fromAminoMethod) {
    return fromAmino.long(args);
  },
  sint64(args: fromAminoMethod) {
    return fromAmino.long(args);
  },
  fixed64(args: fromAminoMethod) {
    return fromAmino.long(args);
  },
  sfixed64(args: fromAminoMethod) {
    return fromAmino.long(args);
  },

  // message.signDoc = object.signDoc !== undefined && object.signDoc !== null ? SignDocDirectAux.fromAmino(object.signDoc) : SignDocDirectAux.fromAmino({});
  type(args: fromAminoMethod) {
    const { origName } = getFieldNames(args.field);
    let name = args.context.getTypeName(args.field);

    if (
      !args.context.options.aminoEncoding.useLegacyInlineEncoding &&
      args.context.options.interfaces.enabled &&
      args.context.options.interfaces?.useGlobalDecoderRegistry &&
      args.field.type === "google.protobuf.Any" &&
      args.field.options["(cosmos_proto.accepts_interface)"]
    ) {
      name = "GlobalDecoderRegistry";
    }

    return setNotUndefinedAndNotNull(
      t.callExpression(
        t.memberExpression(t.identifier(name), t.identifier("fromAmino")),
        [t.memberExpression(t.identifier("object"), t.identifier(origName))]
      ),
      t.identifier("undefined"),
      args
    );
  },

  // message.mode = object.mode ?? 0;
  enum(args: fromAminoMethod) {
    return setNullishCoalescing(
      args,
      getDefaultTSTypeFromProtoType(args.context, args.field, args.isOneOf)
    );
  },

  // message.queryData = object.queryData ?? new Uint8Array()
  bytes(args: fromAminoMethod) {
    return setNullishCoalescing(
      args,
      getDefaultTSTypeFromProtoType(args.context, args.field, args.isOneOf)
    );
  },

  // message.period = object.period ?? undefined;

  duration(args: fromAminoMethod) {
    const durationFormat = args.context.pluginValue(
      "prototypes.typingsFormat.duration"
    );
    switch (durationFormat) {
      case "string":
        return fromAmino.durationString(args);
      case "duration":
      default:
        return fromAmino.type(args);
    }
  },

  durationString(args: fromAminoMethod) {
    return setNullishCoalescing(args, t.identifier("undefined"));
  },

  timestamp(args: fromAminoMethod) {
    const timestampFormat = args.context.pluginValue(
      "prototypes.typingsFormat.timestamp"
    );
    switch (timestampFormat) {
      case "timestamp":
        return fromAmino.type(args);
      case "date":
      default:
        return fromAmino.timestampDate(args);
    }
  },

  // message.periodReset = object.periodReset ?? undefined;

  timestampDate(args: fromAminoMethod) {
    return setNullishCoalescing(args, t.identifier("undefined"));
  },

  // message.referenceMap = Object.entries(object.referenceMap ?? {}).reduce<{
  //     [key: Long]: Reference;
  //   }>((acc, [key, value]) => {
  //     if (value !== undefined) {
  //       acc[Number(key)] = Reference.fromAmino(value);
  //     }

  //     return acc;
  //   }, {});

  // message.labels = Object.entries(object.typeMap ?? {}).reduce<{
  //     [key: string]: string;
  // }>((acc, [key, value]) => {
  //     if (value !== undefined) {
  //         acc[key] = String(value);
  //     }

  //     return acc;
  // }, {});

  keyHash(args: fromAminoMethod) {
    const { propName, origName } = getFieldNames(args.field);
    const keyType = args.field.keyType;
    const valueType = args.field.parsedType.name;

    let fromAminoWithArgs = null;
    // valueTypeType: string for identifier
    let valueTypeType = valueType;
    switch (valueType) {
      case "string":
        fromAminoWithArgs = t.callExpression(t.identifier("String"), [
          t.identifier("value"),
        ]);
        break;
      case "int32":
      case "uint32":
        valueTypeType = "number";
        fromAminoWithArgs = t.callExpression(t.identifier("Number"), [
          t.identifier("value"),
        ]);
        break;
      case "int64":
      case "uint64":
      case "sint64":
      case "fixed64":
      case "sfixed64":
        TypeLong.addUtil(args.context);

        valueTypeType = TypeLong.getPropType(args.context);
        fromAminoWithArgs = TypeLong.getFromValueWithArgs(
          args.context,
          t.identifier("value")
        );
        break;
      default:
        fromAminoWithArgs = t.callExpression(
          t.memberExpression(
            t.identifier(valueType),
            t.identifier("fromAmino")
          ),
          [t.identifier("value")]
        );
    }

    let wrapKey = null;
    let keyTypeType = null;
    switch (keyType) {
      case "string":
        wrapKey = (a) => a;
        keyTypeType = t.tsStringKeyword();
        break;
      case "int64":
      case "uint64":
      case "sint64":
      case "fixed64":
      case "sfixed64":
        wrapKey = (a) => t.callExpression(t.identifier("Number"), [a]);

        TypeLong.addUtil(args.context);

        keyTypeType = t.tsTypeReference(
          TypeLong.getPropIdentifier(args.context)
        );
        break;
      case "int32":
      case "uint32":
        wrapKey = (a) => t.callExpression(t.identifier("Number"), [a]);
        keyTypeType = t.tsNumberKeyword();
        break;
      default:
        throw new Error("keyHash requires new type. Ask maintainers.");
    }

    return t.expressionStatement(
      t.assignmentExpression(
        "=",
        t.memberExpression(t.identifier("message"), t.identifier(propName)),
        callExpression(
          t.memberExpression(
            t.callExpression(
              t.memberExpression(
                t.identifier("Object"),
                t.identifier("entries")
              ),
              [
                t.logicalExpression(
                  "??",
                  t.memberExpression(
                    t.identifier("object"),
                    t.identifier(origName)
                  ),
                  t.objectExpression([])
                ),
              ]
            ),
            t.identifier("reduce")
          ),
          [
            t.arrowFunctionExpression(
              [
                t.identifier("acc"),
                t.arrayPattern([t.identifier("key"), t.identifier("value")]),
              ],
              t.blockStatement([
                t.ifStatement(
                  t.binaryExpression(
                    "!==",
                    t.identifier("value"),
                    t.identifier("undefined")
                  ),
                  t.blockStatement([
                    t.expressionStatement(
                      t.assignmentExpression(
                        "=",
                        t.memberExpression(
                          t.identifier("acc"),
                          wrapKey(t.identifier("key")),
                          true
                        ),
                        fromAminoWithArgs
                      )
                    ),
                  ])
                ),
                t.returnStatement(t.identifier("acc")),
              ])
            ),
            t.objectExpression([]),
          ],
          t.tsTypeParameterInstantiation([
            t.tsTypeLiteral([
              t.tsIndexSignature(
                [identifier("key", t.tsTypeAnnotation(keyTypeType))],
                t.tsTypeAnnotation(
                  t.tsTypeReference(t.identifier(valueTypeType))
                )
              ),
            ]),
          ])
        )
      )
    );
  },

  // message.codeIds = object.codeIds?.map(e => Long.fromValue(e)) || [];
  array(args: fromAminoMethod, expr: t.Expression) {
    const { origName, propName } = getFieldNames(args.field);
    return t.expressionStatement(
      t.assignmentExpression(
        "=",
        t.memberExpression(t.identifier("message"), t.identifier(propName)),
        t.logicalExpression(
          "||",
          t.optionalCallExpression(
            t.optionalMemberExpression(
              t.memberExpression(t.identifier("object"), t.identifier(origName)),
              t.identifier("map"),
              false,
              true
            ),
            [t.arrowFunctionExpression([t.identifier("e")], expr)],
            false
          ),
          t.arrayExpression([])
        )
      )
    );
  },
};

export const arrayTypes = {
  // message.overloadId = object.overloadId?.map(e => e) || [];
  identity() {
    return t.identifier("e");
  },
  string() {
    return arrayTypes.identity();
  },
  bool() {
    return arrayTypes.identity();
  },
  bytes() {
    return arrayTypes.identity();
  },
  double() {
    return arrayTypes.identity();
  },
  float() {
    return arrayTypes.identity();
  },
  int32() {
    return arrayTypes.identity();
  },
  uint32() {
    return arrayTypes.identity();
  },
  sint32() {
    return arrayTypes.identity();
  },
  fixed32() {
    return arrayTypes.identity();
  },
  sfixed32() {
    return arrayTypes.identity();
  },
  enum() {
    return arrayTypes.identity();
  },

  // message.codeIds = object.codeIds?.map(e => Long.fromValue(e)) || [];
  long(args: fromAminoMethod) {
    TypeLong.addUtil(args.context);

    return TypeLong.getFromValueWithArgs(args.context, t.identifier("e"));
  },
  int64(args: fromAminoMethod) {
    return arrayTypes.long(args);
  },
  uint64(args: fromAminoMethod) {
    return arrayTypes.long(args);
  },
  sint64(args: fromAminoMethod) {
    return arrayTypes.long(args);
  },
  fixed64(args: fromAminoMethod) {
    return arrayTypes.long(args);
  },
  sfixed64(args: fromAminoMethod) {
    return arrayTypes.long(args);
  },

  // message.tokenInMaxs = object.tokenInMaxs?.map(e => Coin.fromAmino(e)) || [];
  type(args: fromAminoMethod) {
    let name = args.context.getTypeName(args.field);

    if (
      !args.context.options.aminoEncoding.useLegacyInlineEncoding &&
      args.context.options.interfaces.enabled &&
      args.context.options.interfaces?.useGlobalDecoderRegistry &&
      args.field.type === "google.protobuf.Any" &&
      args.field.options["(cosmos_proto.accepts_interface)"]
    ) {
      name = "GlobalDecoderRegistry";
    }

    return t.callExpression(
      t.memberExpression(t.identifier(name), t.identifier("fromAmino")),
      [t.identifier("e")]
    );
  },
};

export const fromAminoMessages = {
  anyType() {
    return [
      t.returnStatement(
        t.objectExpression([
          t.objectProperty(
            t.identifier("typeUrl"),
            t.memberExpression(t.identifier("object"), t.identifier("type"))
          ),
          t.objectProperty(
            t.identifier("value"),
            t.memberExpression(t.identifier("object"), t.identifier("value"))
          ),
        ])
      ),
    ];
  },
  timestamp(context: ProtoParseContext, name: string, proto: ProtoType) {
    context.addUtil("fromJsonTimestamp");

    return [
      t.returnStatement(
        t.callExpression(t.identifier("fromJsonTimestamp"), [
          t.identifier("object"),
        ])
      ),
    ];
  },
  height(context: ProtoParseContext, name: string, proto: ProtoType) {
    TypeLong.addUtil(context);

    const keepCase = context.options.prototypes.parser.keepCase;
    const casing = keepCase ? (str) => str : camel;

    const makeArgs = (fieldName: string) => {
      let args: t.Expression[] = [
        t.logicalExpression(
          "||",
          t.memberExpression(t.identifier("object"), t.identifier(fieldName)),
          t.stringLiteral("0")
        ),
      ];

      const longType = TypeLong.getType(context);

      if (longType == "Long") {
        args.push(t.booleanLiteral(true));
      }

      return args;
    };

    const makeField = (fieldName: string) =>
      t.objectProperty(
        t.identifier(casing(fieldName)),
        t.callExpression(TypeLong.getFromString(context), makeArgs(fieldName))
      );

    return [
      // return
      t.returnStatement(
        t.objectExpression([
          makeField("revision_number"),
          makeField("revision_height"),
        ])
      ),
    ];
  },
  duration(context: ProtoParseContext, name: string, proto: ProtoType) {
    TypeLong.addUtil(context);

    const longType = TypeLong.getType(context);

    switch (longType) {
      case "BigInt":
        return [
          t.variableDeclaration("const", [
            t.variableDeclarator(
              t.identifier("value"),
              t.callExpression(t.identifier("BigInt"), [t.identifier("object")])
            ),
          ]),
          t.returnStatement(
            t.objectExpression([
              t.objectProperty(
                t.identifier("seconds"),
                t.binaryExpression(
                  "/",
                  t.identifier("value"),
                  t.callExpression(t.identifier("BigInt"), [
                    t.stringLiteral("1000000000"),
                  ])
                )
              ),
              t.objectProperty(
                t.identifier("nanos"),
                t.callExpression(t.identifier("Number"), [
                  t.binaryExpression(
                    "%",
                    t.identifier("value"),
                    t.callExpression(t.identifier("BigInt"), [
                      t.stringLiteral("1000000000"),
                    ])
                  ),
                ])
              ),
            ])
          ),
        ];

      case "Long":
      default:
        return [
          t.variableDeclaration("const", [
            t.variableDeclarator(
              t.identifier("value"),
              t.callExpression(t.identifier("parseInt"), [
                t.identifier("object"),
              ])
            ),
          ]),
          // return
          t.returnStatement(
            t.objectExpression([
              // seconds
              t.objectProperty(
                t.identifier("seconds"),
                t.callExpression(TypeLong.getFromNumber(context), [
                  t.callExpression(
                    t.memberExpression(
                      t.identifier("Math"),
                      t.identifier("floor")
                    ),
                    [t.binaryExpression("/", t.identifier("value"), BILLION)]
                  ),
                ])
              ),

              // nanos
              t.objectProperty(
                t.identifier("nanos"),
                t.binaryExpression("%", t.identifier("value"), BILLION)
              ),
            ])
          ),
        ];
    }
  },
};
