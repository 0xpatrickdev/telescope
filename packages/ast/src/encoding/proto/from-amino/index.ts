import * as t from '@babel/types';
import { ProtoType, ProtoField } from '@osmonauts/types';
import { pascal } from 'case';
import { getFieldOptionality, getFieldOptionalityForDefaults, getOneOfs } from '..';
import { identifier, objectMethod } from '../../../utils';
import { ProtoParseContext } from '../../context';
import { fromAminoJSON, arrayTypes } from './utils';

const needsImplementation = (name: string, field: ProtoField) => {
    throw new Error(`need to implement fromAminoJSON (${field.type} rules[${field.rule}] name[${name}])`);
}
export interface FromAminoJSONMethod {
    context: ProtoParseContext;
    field: ProtoField;
    isOptional: boolean;
}

export const fromAminoJSONMethodFields = (context: ProtoParseContext, name: string, proto: ProtoType) => {
    const oneOfs = getOneOfs(proto);
    const fields = Object.keys(proto.fields ?? {}).map(fieldName => {
        const field = {
            name: fieldName,
            ...proto.fields[fieldName]
        };

        const isOneOf = oneOfs.includes(fieldName);
        const isOptional = getFieldOptionalityForDefaults(context, field, isOneOf);

        const args: FromAminoJSONMethod = {
            context,
            field,
            isOptional
        };

        if (field.rule === 'repeated') {
            switch (field.type) {
                case 'string':
                    return fromAminoJSON.array(args, arrayTypes.string());
                case 'bytes':
                    return fromAminoJSON.array(args, arrayTypes.bytes(args));
                case 'bool':
                    return fromAminoJSON.array(args, arrayTypes.bool());
                case 'float':
                    return fromAminoJSON.array(args, arrayTypes.float());
                case 'double':
                    return fromAminoJSON.array(args, arrayTypes.double());
                case 'int32':
                    return fromAminoJSON.array(args, arrayTypes.int32());
                case 'sint32':
                    return fromAminoJSON.array(args, arrayTypes.sint32());
                case 'uint32':
                    return fromAminoJSON.array(args, arrayTypes.uint32());
                case 'fixed32':
                    return fromAminoJSON.array(args, arrayTypes.fixed32());
                case 'sfixed32':
                    return fromAminoJSON.array(args, arrayTypes.sfixed32());
                case 'int64':
                    return fromAminoJSON.array(args, arrayTypes.int64());
                case 'sint64':
                    return fromAminoJSON.array(args, arrayTypes.sint64());
                case 'uint64':
                    return fromAminoJSON.array(args, arrayTypes.uint64());
                case 'fixed64':
                    return fromAminoJSON.array(args, arrayTypes.fixed64());
                case 'sfixed64':
                    return fromAminoJSON.array(args, arrayTypes.sfixed64());
                default:
                    switch (field.parsedType.type) {
                        case 'Enum':
                            return fromAminoJSON.array(args, arrayTypes.enum(args));
                        case 'Type':
                            return fromAminoJSON.array(args, arrayTypes.type(args));
                    }
                    return needsImplementation(fieldName, field);
            }
        }

        if (field.keyType) {
            switch (field.keyType) {
                case 'string':
                case 'int64':
                case 'sint64':
                case 'uint64':
                case 'fixed64':
                case 'sfixed64':
                case 'int32':
                case 'sint32':
                case 'uint32':
                case 'fixed32':
                case 'sfixed32':
                    return fromAminoJSON.keyHash(args);
                default:
                    return needsImplementation(fieldName, field);
            }
        }

        switch (field.type) {
            case 'string':
                return fromAminoJSON.string(args);
            case 'bytes':
                return fromAminoJSON.bytes(args);
            case 'bool':
                return fromAminoJSON.bool(args);
            case 'double':
                return fromAminoJSON.double(args);
            case 'float':
                return fromAminoJSON.float(args);
            case 'int32':
                return fromAminoJSON.int32(args);
            case 'sint32':
                return fromAminoJSON.sint32(args);
            case 'uint32':
                return fromAminoJSON.uint32(args);
            case 'fixed32':
                return fromAminoJSON.fixed32(args);
            case 'sfixed32':
                return fromAminoJSON.sfixed32(args);
            case 'int64':
                return fromAminoJSON.int64(args);
            case 'sint64':
                return fromAminoJSON.sint64(args);
            case 'uint64':
                return fromAminoJSON.uint64(args);
            case 'fixed64':
                return fromAminoJSON.fixed64(args);
            case 'sfixed64':
                return fromAminoJSON.sfixed64(args);
            case 'Duration':
            case 'google.protobuf.Duration':
                return fromAminoJSON.duration(args);
            case 'Timestamp':
            case 'google.protobuf.Timestamp':
                return fromAminoJSON.timestamp(args);
            default:
                switch (field.parsedType.type) {
                    case 'Enum':
                        return fromAminoJSON.enum(args);
                    case 'Type':
                        return fromAminoJSON.type(args);
                }
                return needsImplementation(fieldName, field);
        }
    });
    return fields;
};


export const fromAminoJSONMethod = (context: ProtoParseContext, name: string, proto: ProtoType) => {
    const fields = fromAminoJSONMethodFields(context, name, proto);
    let varName = 'object';
    if (!fields.length) {
        varName = '_';
    }

    const SDKTypeName =
        [name, 'SDKType']
            .filter(Boolean).join('');

    if (proto.type === 'Type' && proto.name === 'Duration') {
        // console.log(proto);
        // we could do something special here....
    }

    return objectMethod('method',
        t.identifier('fromAmino'),
        [
            identifier(varName,
                t.tsTypeAnnotation(
                    t.tsTypeReference(
                        t.identifier(SDKTypeName)
                    )
                ),
                false
            )

        ],
        t.blockStatement(
            [
                // t.variableDeclaration('const', [
                //     t.variableDeclarator(
                //         t.identifier('hoistedAsts'),
                //         t.nullLiteral()
                //     )
                // ]),

                t.returnStatement(
                    t.objectExpression(fields)
                )
            ]
        ),
        false,
        false,
        false,
        t.tsTypeAnnotation(
            t.tsTypeReference(
                t.identifier(name)
            )
        )
    )
};
