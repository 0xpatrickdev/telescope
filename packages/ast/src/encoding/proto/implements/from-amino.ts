import * as t from '@babel/types';
import { arrowFunctionExpression, identifier } from '../../../utils';
import { ProtoParseContext } from "../../context";

export const createInterfaceFromAminoHelper = (
    context: ProtoParseContext,
    functionName: string
) => {

    const switchCases = [
        t.switchCase(
            t.stringLiteral('cosmos-sdk/TextProposal'),
            [
                t.returnStatement(
                    t.callExpression(
                        t.memberExpression(
                            t.identifier('Any'),
                            t.identifier('fromPartial')
                        ),
                        [
                            t.objectExpression([
                                t.objectProperty(
                                    t.identifier('typeUrl'),
                                    t.stringLiteral('/cosmos.gov.v1beta1.TextProposal')
                                ),
                                t.objectProperty(
                                    t.identifier('value'),
                                    t.callExpression(
                                        t.memberExpression(
                                            t.callExpression(
                                                t.memberExpression(
                                                    t.identifier('TextProposal'),
                                                    t.identifier('encode')
                                                ),
                                                [
                                                    t.callExpression(
                                                        t.memberExpression(
                                                            t.identifier('TextProposal'),
                                                            t.identifier('fromPartial')
                                                        ),
                                                        [
                                                            t.tsAsExpression(
                                                                t.memberExpression(
                                                                    t.identifier('content'),
                                                                    t.identifier('value')
                                                                ),
                                                                t.tsTypeReference(
                                                                    t.identifier('DeepPartial'),
                                                                    t.tsTypeParameterInstantiation([
                                                                        t.tsTypeReference(
                                                                            t.identifier('TextProposal')
                                                                        )
                                                                    ])
                                                                )
                                                            )
                                                        ]
                                                    )
                                                ]
                                            ),
                                            t.identifier('finish')
                                        ),
                                        []
                                    )
                                ),

                            ])
                        ]
                    )
                )]
        )
    ];

    return t.exportNamedDeclaration(
        t.variableDeclaration(
            'const',
            [
                t.variableDeclarator(
                    t.identifier(functionName),
                    t.arrowFunctionExpression(
                        [
                            // params
                        ],
                        t.blockStatement([
                            t.switchStatement(
                                t.memberExpression(
                                    t.identifier('content'),
                                    t.identifier('type')
                                ),
                                [
                                    ...switchCases,
                                    /// default
                                    t.switchCase(
                                        null,
                                        [
                                            t.returnStatement(
                                                t.callExpression(
                                                    t.memberExpression(
                                                        t.identifier('Any'),
                                                        t.identifier('fromAmino')
                                                    ),
                                                    [
                                                        t.identifier('content')
                                                    ]
                                                )
                                            )
                                        ]
                                    )
                                ]
                            )
                        ])
                    )
                )
            ]
        )
    )
};