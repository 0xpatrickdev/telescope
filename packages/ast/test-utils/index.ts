import generate from '@babel/generator';
import { ProtoStore, getNestedProto, parseProto } from '@osmonauts/proto-parser'
import { ProtoType } from '../src/encoding/proto/types';
import { ParseContext } from '../src/encoding/context';

export const expectCode = (ast) => {
    expect(
        generate(ast).code
    ).toMatchSnapshot();
}
export const printCode = (ast) => {
    console.log(
        generate(ast).code
    );
}

export const prepareContext = (store: ProtoStore, protoFile: string) => {
    const ref = store.findProto(protoFile);
    const traversed = ref.traversed
    if (!traversed) {
        throw new Error('not traversed!');
    }
    const proto: Record<string, ProtoType> = getNestedProto(traversed);
    const protos: ProtoType[] = Object.values(proto).filter(
        proto => proto.name.startsWith('Msg')
            &&
            !proto.name.endsWith('Response')
            &&
            proto.name !== 'Msg'
    );

    const context = new ParseContext(ref, store);

    return {
        context,
        root: traversed,
        protos
    };
};
