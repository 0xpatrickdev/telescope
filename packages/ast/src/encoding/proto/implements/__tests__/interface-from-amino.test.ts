import { expectCode, printCode, getTestProtoStore } from '../../../../../test-utils/'
import { ProtoParseContext } from '../../../context';
import { createInterfaceFromAminoHelper } from '../from-amino';

const store = getTestProtoStore();
store.options.prototypes.implementsAcceptsAny = true;
store.options.aminoEncoding.enabled = true;
store.options.aminoEncoding.useRecursiveV2encoding = true;
store.traverseAll();

describe('ProposalContent', () => {
    const txRef = store.findProto('cosmos/gov/v1/tx.proto');
    const txContext = new ProtoParseContext(txRef, store, store.options);
    it('ProposalContent', () => {
        expectCode(createInterfaceFromAminoHelper(
            txContext,
            'ProposalContentI_fromAmino'
        ));
    });
});