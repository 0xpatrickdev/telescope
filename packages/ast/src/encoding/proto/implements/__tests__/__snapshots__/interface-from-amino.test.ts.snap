// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProposalContent ProposalContent 1`] = `
"export const ProposalContentI_fromAmino = () => {
  switch (content.type) {
    case "cosmos-sdk/TextProposal":
      return Any.fromPartial({
        typeUrl: "/cosmos.gov.v1beta1.TextProposal",
        value: TextProposal.encode(TextProposal.fromPartial((content.value as DeepPartial<TextProposal>))).finish()
      });

    default:
      return Any.fromAmino(content);
  }
};"
`;
